content/specs/auth.md

```diff
@@ -0,0 +1,389 @@
+
+To make development workflows easier, a special exception is made for clients with `client_id` having origin `http://localhost` (with no port number specified). Authorization Servers are encouraged to support this exception - including in production environments - but it is optional.
+
+In a localhost `client_id` scenario, the Authorization Server should verify that the scheme is `http`, and that the hostname is exactly `localhost` with no port specified. IP addresses (`127.0.0.1`, etc) are not supported. The path parameter must be empty (`/`). Some metadata fields can be configured via query parameter in the `client_id` URL (with appropriate urlencoding):
+
+- `redirect_uri` (string, multiple allowed, optional): allows declaring a local redirect/callback URL. Otherwise, `http://127.0.0.1/` and `http://[::1]/` will be used.
+- `scope` (string, space-separated, multiple allowed, optional): the set of scopes which might be requested by the client. Default is `atproto`.
+
+The other parameters in the virtual client metadata document will be:
+
+- `client_id` (string): the exact `client_id` (URL) used to generate the virtual document
+- `client_name` (string): a value chosen by the Authorization Server (e.g. "Loopback client")
+- `response_types` (array of strings): must include `code`
+- `grant_types` (array of strings): `authorization_code` and `refresh_token`
+- `token_endpoint_auth_method`: `none`
+- `application_type`: `native`
+- `dpop_bound_access_tokens`: `true`
+
+Note that this works as a public client, not a confidential client.
+
+### Identity Authentication
+
+As mentioned in the introduction, OAuth 2.0 generally provides only Authorization (`authz`), and additional standards like OpenID/OIDC are used for Authentication (`authn`). The atproto profile of OAuth requires authentication of account identity and supports the use case of simple identity authentication without additional resource access authorization.
+
+In atproto, account identity is anchored in the account DID, which is the permanent, globally unique, publicly resolvable identifier for the account. The DID resolves to a DID document which indicates the current PDS host location for the account. That PDS (combined with an optional entryway) is the authorization authority and the OAuth Authorization Server for the account. When speaking to any Authorization Server, it is critical (mandatory) for clients to confirm that it is actually the authoritative server for the account in question, which means independently resolving the account identity (by DID) and confirming that the Authorization Server matches. It is also critical (mandatory) to confirm at the end of an authorization flow that the Authorization Server actually granted tokens for the expected account.
```

Comment:
Use RFC language here.

> When speaking to any Authorization Server clients _must_ confirm that the authorization server is the authoritative for the account being authorized. Clients _must_ do this by resolving the account identity to a DID, resolving the PDS for the DID and finally resolving the authorisation server for that PDS. (See appendix _<whatever>_ for further details)

This however I disagree with

> It is also critical (mandatory) to confirm at the end of an authorization flow that the Authorization Server actually granted tokens for the expected account.

It is not the client's job to validate tokens given to it by an authorization server. That is the job of the server who requires the tokens. The oauth spec does not specify a token format (although [rfc9068](https://datatracker.ietf.org/doc/html/rfc9068) does, but it's in draft, and no-one appears to have picked it up again since the author passed away). 

It's spec valid to get a guid back, or a reference token instead of JWTs. Clients should be handling tokens as opaque and just passing them onto the api server. (Although I realise in practice people do look at expiry dates etc. when it's a JWT, but what happens then if you switch to encrypted JWTS? They're encrypted for the server, and the client can't do anything with them at all)

Reply:
I'm sure we'll do a whole pass over these specs to use RFC 2119 terminology (maybe when these actually get written to submit as I-D), but that hasn't happened yet.

I think the end of this paragraph was confusing so I updated it. We aren't asking clients to inspect the tokens in any way (they are opaque to the client). What they need to inspect is the `sub` field of the token response body (which includes the tokens themselves in separate fields).